source/idf/EthernetDevice.cpp

```diff
@@ -0,0 +1,162 @@
+#include "idf/EthernetDevice.hh"
+#include "idf/IOException.hh"
+
+#include <errno.h>
+#include <cstring>
+#include <iostream>
+#include <sstream>
+#include <netinet/in.h>
+#include <arpa/inet.h>
+#include <sys/socket.h>
+#include <unistd.h>
+
+namespace idf {
+
+EthernetDevice::EthernetDevice(const std::string& id, unsigned length) :
+    InputDevice(id),
+    packetLength(length),
+    sockType(SOCK_STREAM),
+    tcp(true) {
+        memset(&serverAddr, 0 , sizeof(serverAddr));
+        setTCP(); // ensure vars are zeroed out.
```

Comment:
hmmmm hadn't considered that. It was just done out of convenience to default to TCP and init accordingly. An inheriting device might need some alternative TCP init... Advise?

Reply:
hmmm... maybe then not default to TCP. Throw an IOException if attempt to open and TCP or UDP isn't set.

Reply:
one way to handle it could be to make a `setTCPBase` method that isn't virtual and can be called safely in the constructor. if the user wants to change the behavior of the constructor, they would implement it in their derived-classes constructor which would act upon the instance knowing setTCPBase was called already.